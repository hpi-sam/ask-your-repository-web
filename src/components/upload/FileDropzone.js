// @flow
import React from 'react';
import Dropzone from 'react-dropzone';
import type { DropFilesEventHandler } from 'react-dropzone';
import './FileDropzone.scss';

type Props = {
  onDrop: DropFilesEventHandler,
};

const FileDropzone = ({ onDrop }: Props) => (
  <Dropzone
    className="FileDropzone"
    activeClassName="FileDropzone--active"
    acceptClassName="FileDropzone--accept"
    disabledClassName="FileDropzone--disabled"
    onDrop={onDrop}
    data-cy="upload-dropzone"
  >
    <span className="FileDropzone__text">
      Place your file(s) here
    </span>
  </Dropzone>
);

export default FileDropzone;
