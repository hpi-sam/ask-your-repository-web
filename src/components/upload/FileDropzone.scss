@import '../../style/breakpoints.scss';

.FileDropzone {
  --bg-color: transparent;
  --hover-bg-color: rgba(255, 255, 255, .03);

  --border-width: 4px;
  --border-color: #fff;
  --hover-border-color: rgb(36, 186, 255);

  --text-color: #fff;
  --hover-text-color: rgb(36, 186, 255);

  --transition-duration: 0.2s;

  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;

  height: 100%;
  width: 100%;
  padding: 20px;

  background-color: var(--bg-color);
  border: var(--border-width) dashed var(--border-color);
  border-radius: inherit;

  transition:
    background-color var(--transition-duration),
    border-color var(--transition-duration),
    color var(--transition-duration);

  &:hover {
    --bg-color: var(--hover-bg-color);
    --text-color: var(--hover-text-color);
    --border-color: var(--hover-border-color);
  }

  &__text {
    text-align: center;
    font-weight: bold;
    letter-spacing: 3px;
    text-transform: uppercase;
    color: var(--text-color);
  }
}
