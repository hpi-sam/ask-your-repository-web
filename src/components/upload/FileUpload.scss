@import '../../style/breakpoints.scss';

.FileUpload {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  height: 90vh;
  width: 100%;
  padding: 30px;

  &__inner {
    display: flex;
    align-items: center;
    justify-content: center;
    border-radius: 10px;
    background-color: rgba(0, 0, 0, .2);
    height: 90vh;
    width: 100%;

    @media (min-width: $xs-to-sm) {
      width: 750px;
      height: 300px;
    }
  }

}
