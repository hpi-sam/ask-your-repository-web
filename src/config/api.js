import axios from 'axios';
import qs from 'qs';

const api = axios.create({
  baseURL: process.env.REACT_APP_API_URL,
  requestInterceptor: config => ({
    ...config,
  }),
  paramsSerializer: params => qs.stringify(params, { arrayFormat: 'repeat' }),
  withCredentials: true,
});

export default api;
