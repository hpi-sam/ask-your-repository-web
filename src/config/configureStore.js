// @flow
import { applyMiddleware, createStore } from 'redux';
import type { Dispatch } from 'redux';
import { persistStore, persistReducer } from 'redux-persist-erksch';
import storage from 'redux-persist-erksch/lib/storage';
import { routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { middleware as flashMiddleware } from 'redux-flash';
import thunk from 'redux-thunk';
import socketioMiddleware from '../middleware/socketioMiddleware';
import axiosAuthMiddleware from './axiosAuthMiddleware';
import createRootReducer from '../state/rootReducer';
import type { AppState } from '../state/AppState';
import type { Action } from '../state/Action';

export const history = createBrowserHistory();

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['activeTeam', 'presentationMode', 'auth'],
};

function configureStore() {
  let middleware = [
    thunk,
    flashMiddleware(),
    routerMiddleware(history),
    socketioMiddleware(),
    axiosAuthMiddleware(),
  ];

  if (process.env.NODE_ENV === 'development') {
    const { logger } = require('redux-logger');
    middleware = [...middleware, logger];
  }

  const rootReducer = createRootReducer(history);
  const persistedReducer: any = persistReducer(persistConfig, rootReducer);

  const store = createStore<AppState, Action, Dispatch<Action>>(
    persistedReducer,
    applyMiddleware(...middleware),
  );

  const persistor = persistStore(store);

  return { store, persistor };
}

export default configureStore;
