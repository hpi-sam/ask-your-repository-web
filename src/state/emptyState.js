// @flow
import type { AppState } from './AppState';

const emptyState: AppState = {
  flash: { messages: [] },
  image: null,
};

export default emptyState;
