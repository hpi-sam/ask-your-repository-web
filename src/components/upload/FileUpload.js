// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { flashErrorMessage } from 'redux-flash';
import { Redirect } from 'react-router-dom';
import FileDropzone from './FileDropzone';
import ActivityIndicator from '../utility/ActivityIndicator';
import fetchUploadImage from '../../requests/imageRequests';
import { setImage } from '../../state/image/image.actionCreators';
import './FileUpload.scss';

type Props = {
  dispatch: Function,
};

type State = {
  imageId: ?number,
  isUploading: boolean,
  hasUploaded: boolean,
};

class FileUpload extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      imageId: null,
      isUploading: false,
      hasUploaded: false,
    };
  }

  setImageState(dataUri: string) {
    const { imageId } = this.state;
    if (!imageId) return;

    this.props.dispatch(setImage(imageId, dataUri));
  }

  handleImageDrop = async (files: Blob[]) => {
    if (files && files[0]) {
      const image = files[0];
      await this.submitImage(image);

      if (this.state.hasUploaded) {
        const reader = new FileReader();
        reader.onload = ({ target }) => this.setImageState(target.result);
        reader.readAsDataURL(image);
      }
    }
  };

  async submitImage(image: Blob) {
    const formData = new FormData();
    formData.append('image', image);

    this.setState({ isUploading: true });

    await fetchUploadImage(formData)
      .then(({ data }) => {
        this.setState({
          hasUploaded: true,
          imageId: data.id,
        });
      })
      .catch((error) => {
        const message = error.response
          ? error.response.message
          : 'Our server seems to be unavailable at the moment.';
        this.props.dispatch(flashErrorMessage(message));
      });

    this.setState({ isUploading: false });
  }

  render() {
    if (this.state.hasUploaded) {
      return <Redirect to="/tagging" />;
    }

    const { isUploading } = this.state;

    return (
      <div className="FileUpload">
        <div className="FileUpload__inner">
          {isUploading ? (
            <ActivityIndicator text="Uploading" />
          ) : (
            <FileDropzone onDrop={this.handleImageDrop} />
          )}
        </div>
      </div>
    );
  }
}

export default connect()(FileUpload);
