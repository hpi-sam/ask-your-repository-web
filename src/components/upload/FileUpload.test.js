// @flow
import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import FileUpload from './FileUpload';
import FileDropzone from './FileDropzone';
import api from '../../config/api';
import emptyState from '../../state/emptyState';

const mockStore = configureStore();

jest.mock('../../config/api', () => {
  const responseData = { id: 1 };

  return {
    post: jest.fn(() => Promise.resolve({ data: responseData })),
  };
});

describe('<FileUpload />', () => {
  let wrapper;
  let wrapperInstance;
  let store;

  beforeEach(() => {
    store = mockStore(emptyState);
    wrapper = shallow(<FileUpload store={store} />).dive();
    wrapperInstance = wrapper.instance();
  });

  afterEach(() => {
    api.post.mockClear();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders an FileDropzone component', () => {
    expect(wrapper.find(FileDropzone).exists()).toBeTruthy();
  });

  describe('an image has been dropped', () => {
    const imageBlob = new Blob([], { type: 'image/jpeg' });

    beforeEach(async () => {
      await wrapperInstance.handleImageDrop([imageBlob]);
    });

    it('should send a POST request to the server', () => {
      expect(api.post).toHaveBeenCalled();
    });

    it('should set the correct Content-Type for the request', () => {
      const { headers } = api.post.mock.calls[0][2];
      expect(headers['Content-Type']).toEqual('multipart/form-data');
    });

    it('should send the dropped file as form-data', async () => {
      const formData = api.post.mock.calls[0][1];
      const sentFile = formData.get('image');

      let sentFileData: string;
      let imageBlobData: string;

      function asyncFileReader(file: Blob, onLoad: (string) => void) {
        return new Promise((resolve) => {
          const fileReader = new FileReader();
          fileReader.onload = (e) => {
            onLoad(e.target.result);
            resolve();
          };
          fileReader.readAsDataURL(file);
        });
      }

      await Promise.all([
        asyncFileReader(sentFile, (result) => { sentFileData = result; }),
        asyncFileReader(imageBlob, (result) => { imageBlobData = result; }),
      ]);

      expect(sentFileData).toEqual(imageBlobData);
    });

    describe('upload failed because of network error', () => {
      beforeAll(() => {
        api.post.mockImplementation(() => Promise.reject(Response.error()));
      });

      it('should dispatch a network error message flash message', () => {
        expect(store.getActions()[0].payload).toEqual(
          expect.objectContaining({
            options: { isError: true },
            message: expect.stringContaining('unavailable'),
          }),
        );
      });
    });

    describe('upload failed because of server error', () => {
      const errorMessage = 'Dummy Error';
      const errorResponse = { message: errorMessage };

      beforeAll(() => {
        api.post.mockImplementation(() => Promise.reject({ response: errorResponse }));
      });

      it('should dispatch an error flash message', () => {
        expect(store.getActions()[0]).toEqual(
          expect.objectContaining({
            payload: {
              options: { isError: true },
              message: errorMessage,
            },
          }),
        );
      });
    });
  });

  it('do nothing if image drop is called without any files', () => {
    wrapperInstance.handleImageDrop(null);
    expect(api.post).not.toHaveBeenCalled();
  });
});
