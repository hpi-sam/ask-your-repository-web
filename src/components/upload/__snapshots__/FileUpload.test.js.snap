// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileUpload /> renders correctly 1`] = `
<div
  className="FileUpload"
>
  <div
    className="FileUpload__inner"
  >
    <FileDropzone
      onDrop={[Function]}
    />
  </div>
</div>
`;
